### DISCUSSION TITLE
Strategies for reducing/eliminating module duplication across chunks

### DISCUSSION THREAD

**Discussion by scameron** [webpack/webpack#19024] [General]
I've never been able to understand the exact reason why, but I often see modules duplicated across multiple chunks in our build.  I know splitChunks tries to split out common modules into separate chunks that get shared between chunk groups, but for reasons that are mysterious to me, it doesn't always do this.

Typically, the duplicated modules are modules that are statically imported by code in multiple async loaded chunk groups.  Not all of the common modules are duplicated into each chunk group, but a significant number are.

If we assume in a complex enough codebase splitChunks can't automatically do a perfect job of deduping, what else can be done?  I've played around a bit with cache groups but I've never been able to get a satisfactory outcome.  Explicitly referencing individual modules seems brittle because code changes over time.  And pulling in entire folders or directories can end up adding too much to a cache group, subverting the normal build optimizations.  Maybe there's a way to find the sweet spot here, but I haven't found it yet.

Does anybody have any good strategies for analyzing duplication and eliminating it?  If anybody has been able to get a good outcome in a reasonably large/complex code base, I'd be super interested in any ideas or hints you might have in mind.

Thanks in advance!

**Comment by alexander-akait** [comment]
Hi, this is actually quite a complex task, we have many parameters to add a better spliting.

I would recommend looking at:
- https://webpack.js.org/plugins/split-chunks-plugin/#splitchunksminremainingsize
- https://webpack.js.org/plugins/split-chunks-plugin/#splitchunkscachegroupscachegroupreuseexistingchunk

**Reply by scameron** [reply]
The biggest problem for me is that I have no conceptual model at all for how to this.  I've read everything I can find about cache groups and tried to understand the chunk splitting code, but I still don't have a clear idea about exactly how cache groups concretely influence chunk splitting.  Without that level of understanding, I feel like I'm just tweaking things at random.  I have no way to reason about the choices.

The one thing in cache groups I do understand is the following:  if I define `name` as a constant string, then all modules matched by the cache group settings will be added into a single named chunk and further splitting is disabled.  That works and is clearly observable, but it subverts the chunk splitting algorithm, so I don't want to use that too much.

There are 2 default cache groups: `default` and `defaultVendors`.  If I disable these, my build gets bigger and duplication increases.  But why?  How do these cache groups actually influence the chunk splitting?  Clearly defining a separate group for `node_modules` does something, but I have no idea what.

For me, this "magical" chunk splitting is by far the most frustrating aspect of webpack.  Any clarity you can offer would be hugely appreciated.

**Reply by scameron** [reply]
Here's an example of what I mean, fiddling only with the default cache groups and nothing else (all comments are relative to the baseline of just leaving the defaults as they are)

- Disable both `default` and `defaultVendors` cache groups --> build more than doubles in size with 25X more module duplication
- Disable `default` but keep `defaultVendors` alone --> build size increases by 40% and duplication increases by 10%
- Disable `defaultVenders` but keep `default` alone --> build size actually _decreases_ by a tiny bit and duplication decreases by 4%

I tried setting `minChunks: 2` onto `defaultVendors` but it doesn't change much (just a few _more_ duplicated modules).

So it seems that the `default` cache group is beneficial in our case, but the `defaultVendors` cache group either does nothing (if it's together with `default`) or makes things worse (if it's by itself).

I have no idea how to make sense of this information in order to try to determine other cache groups to consider creating for my build...

**Reply by alexander-akait** [reply]
To start, here are the basic parameters we have for groups- https://github.com/webpack/webpack/blob/main/lib/config/defaults.js#L1507

I would advise looking at the options (there are more options, but i want to talk about the most important ones here):
- `reuseExistingChunk`
- `priority`

If the module meets both criteria of the group - 
the optimization will prefer the cache group with a higher priority, but because we use `reuseExistingChunk`, webpack will use the existing chunk.

> So it seems that the default cache group is beneficial in our case, but the defaultVendors cache group either does nothing (if it's together with default) or makes things worse (if it's by itself).

The main task of `defaultVendors` is long-term caching, but not just the optimal splitting. How often do you change `node_modules`? I don't think it's often, so if you will use `[contenthash]` clients may cache some modules (from `node_modules` ) for a very long time. Which will allow you to create fewer requests to the static asset server.

The main task of `splitting` as I said earlier not only in the optimal splitting, but also in the possibility of cache modules into a long cache, so when you think about the optimal splitting, I recommend basing yourself on this assumption too.

Another recommendation is the structure of the application, for example, you have an online store - a catalog, a shopping cart, a personal account (etc), the buyer can't always reach the cart, changing personal information is also not a frequent process, therefore it is logical to allocate functionality for chunks by structure, for example, not to put a personal account logic for the catalog. 

So don't think of splitting as just a set of options (although I would recommend experimenting with them). 

I would like to recommend such things when you think about splitting:
1. Size of chunks.
2. Long term cache.
3. Structure of the application.

Unfortunately, there is really no universal recipe for how to do it better, we offer standard parameters because they satisfy most small and medium applications and allow make a universal splitting without knowledge of structure.

Therefore, before you start changing the parameters of `splitting` ask ask yourself - what do you want to achieve first.
